bundles/org.dataflowanalysis.analysis.pcm/src/org/dataflowanalysis/analysis/pcm/core/PCMVertexCharacteristicsCalculator.java

```diff
@@ -0,0 +1,249 @@
+import org.dataflowanalysis.pcm.extension.nodecharacteristics.nodecharacteristics.UsageAssignee;
+import org.palladiosimulator.pcm.allocation.Allocation;
+import org.palladiosimulator.pcm.allocation.AllocationContext;
+import org.palladiosimulator.pcm.allocation.AllocationPackage;
+import org.palladiosimulator.pcm.core.composition.AssemblyContext;
+import org.palladiosimulator.pcm.core.entity.Entity;
+import org.palladiosimulator.pcm.repository.CompositeComponent;
+import org.palladiosimulator.pcm.repository.Repository;
+import org.palladiosimulator.pcm.repository.RepositoryPackage;
+import org.palladiosimulator.pcm.resourceenvironment.ResourceContainer;
+import org.palladiosimulator.pcm.resourceenvironment.ResourceEnvironment;
+import org.palladiosimulator.pcm.resourceenvironment.ResourceenvironmentPackage;
+import org.palladiosimulator.pcm.seff.AbstractAction;
+import org.palladiosimulator.pcm.system.System;
+import org.palladiosimulator.pcm.system.SystemPackage;
+import org.palladiosimulator.pcm.usagemodel.AbstractUserAction;
+import org.palladiosimulator.pcm.usagemodel.UsageModel;
+import org.palladiosimulator.pcm.usagemodel.UsageScenario;
+import org.palladiosimulator.pcm.usagemodel.UsagemodelPackage;
+
+public class PCMVertexCharacteristicsCalculator {
+    private final Logger logger = Logger.getLogger(PCMVertexCharacteristicsCalculator.class);
+    private final ResourceProvider resourceLoader;
+
+    /**
+     * Creates a new node characteristic calculator with the given resource provider
+     * @param resourceProvider Resource provider that is used to calculate characteristics for each node
+     */
+    public PCMVertexCharacteristicsCalculator(ResourceProvider resourceProvider) {
+        this.resourceLoader = resourceProvider;
+    }
+
+    public List<CharacteristicValue> getNodeCharacteristics(Entity node, Deque<AssemblyContext> context) {
+        Assignments assignments = this.resolveAssignments();
+        List<AbstractAssignee> assignees;
+        if (node instanceof AbstractUserAction) {
+            assignees = this.getUsage(node, assignments);
+        } else if (node instanceof AbstractAction) {
+            assignees = this.getSEFF(assignments, context);
+        } else {
+            throw new IllegalArgumentException("Unknown assignee:" + node.toString());
+        }
+        List<EnumCharacteristic> enumCharacteristics = assignees.stream().flatMap(it -> it.getCharacteristics().stream()).toList();
+
+        return enumCharacteristics.stream().flatMap(it -> it.getValues().stream().map(val -> new PCMCharacteristicValue(it.getType(), val)))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Gets the assignees of a usage node
+     * @param assignments Resolved assignment container
+     * @return List of resolved assignees matching the node
+     */
+    private List<AbstractAssignee> getUsage(Entity node, Assignments assignments) {
```

Comment:
The name of this and the following method does not represent what it actually does (dealing with assignees)

Reply:
Renamed to get{SEFFNode, UserNode, ...} Assignments in 2f73514df5cd9d823803b3aa13b06eacf919c002